cocoa/drawing.m

```diff
@@ -0,0 +1,526 @@
+
+    CFStringRef keys[] = { kCTFontAttributeName, kCTForegroundColorAttributeName };
+    CFTypeRef values[] = { font, global_text_state._use_font_color_ref };
+
+    CFDictionaryRef attributes = CFDictionaryCreate(kCFAllocatorDefault, (const void**)&keys,
+                                                    (const void**)&values, 2,
+                                                    &kCFTypeDictionaryKeyCallBacks,
+                                                    &kCFTypeDictionaryValueCallBacks);
+
+    CFAttributedStringRef attrString = CFAttributedStringCreate(kCFAllocatorDefault, string, attributes);
+    CFRelease(string);
+    CFRelease(attributes);
+
+    CTLineRef line = CTLineCreateWithAttributedString(attrString);
+
+    int ret = round(CTLineGetTypographicBounds(line, NULL, NULL, NULL) + CTLineGetTrailingWhitespaceWidth(line));
+    CFRelease(line);
+    CFRelease(attrString);
+    return ret;
+}
+
+int textfit(char_t *str, STRING_IDX length, int width) {
+    CFStringRef string = CFStringCreateWithBytes(kCFAllocatorDefault, str, length, kCFStringEncodingUTF8, NO);
+    CTFontRef font = global_text_state._use_font;
+
+    CFStringRef keys[] = { kCTFontAttributeName };
+    CFTypeRef values[] = { font };
+
+    CFDictionaryRef attributes = CFDictionaryCreate(kCFAllocatorDefault, (const void**)&keys,
+                                                    (const void**)&values, 1,
+                                                    &kCFTypeDictionaryKeyCallBacks,
+                                                    &kCFTypeDictionaryValueCallBacks);
+
+    CFAttributedStringRef attrString = CFAttributedStringCreate(kCFAllocatorDefault, string, attributes);
+    CFRelease(attributes);
+
+    CTLineRef line = CTLineCreateWithAttributedString(attrString);
+    int ind = CTLineGetStringIndexForPosition(line, (CGPoint){width, 0});
+
+    CFRelease(line);
+    CFRelease(attrString);
+
+    if (ind == -1)
+        ++ind;
+
+    int ret = [[(NSString *)string substringToIndex:ind] lengthOfBytesUsingEncoding:NSUTF8StringEncoding];
+
+    CFRelease(string);
+    return ret;
+}
+
+int textfit_near(char_t *str, STRING_IDX length, int width) {
+    return textfit(str, length, width);
+}
+
+void setfont(int id) {
+    global_text_state._use_font = fonts[id];
+}
+
+uint32_t setcolor(uint32_t color) {
+    uint32_t ret = global_text_state._use_font_color;
+
+    if (global_text_state._use_font_color_ref)
+        CGColorRelease(global_text_state._use_font_color_ref);
+
+    CGColorSpaceRef cs = CGColorSpace_CREATE();
+    float r = ((color >> 16) & 0xFF) / 255.0,
+    g = ((color >> 8) & 0xFF) / 255.0,
+    b = ((color) & 0xFF) / 255.0;
+    CGFloat comp[4] = {r, g, b, 1.0};
+    global_text_state._use_font_color_ref = CGColorCreate(cs, comp);
+    CGColorSpaceRelease(cs);
+
+    global_text_state._use_font_color = color;
+    return ret;
+}
+
+void reload_fonts(void) {
+    for (int i = 0; i < sizeof(fonts) / sizeof(CTFontRef); ++i)
```

Comment:
this too, you probably want to grep regex for `).*\n`

and there's a few above this as well